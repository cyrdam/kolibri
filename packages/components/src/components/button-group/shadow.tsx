import type { ButtonGroupProps } from '../../schema';
import type { JSX } from '@stencil/core';
import { Component, h, Host } from '@stencil/core';
import { KolButtonGroupWcTag } from '../../core/component-names';

@Component({
	tag: 'kol-button-group',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolButtonGroup implements ButtonGroupProps {
	public render(): JSX.Element {
		return (
			<Host class="kol-button-group">
				<KolButtonGroupWcTag>
					<slot />
				</KolButtonGroupWcTag>
			</Host>
		);
	}
}
