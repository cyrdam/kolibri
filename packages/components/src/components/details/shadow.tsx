import { Component, Element, h, type JSX, Method, Prop, State, Watch } from '@stencil/core';
import type { DetailsAPI, DetailsCallbacksPropType, DetailsStates, DisabledPropType, FocusableElement, HeadingLevel, LabelPropType } from '../../schema';
import { validateDetailsCallbacks, validateDisabled, validateLabel, validateOpen } from '../../schema';
import KolCollapsibleFc, { type CollapsibleProps } from '../../functional-components/Collapsible';
import { nonce } from '../../utils/dev.utils';
import { watchHeadingLevel } from '../heading/validation';
import { dispatchDomEvent, KolEvent } from '../../utils/events';

/**
 * @slot - Der Inhalt, der in der Detailbeschreibung angezeigt wird.
 */
@Component({
	tag: 'kol-details',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolDetails implements DetailsAPI, FocusableElement {
	@Element() private readonly host?: HTMLKolDetailsElement;

	private readonly nonce = nonce();
	private buttonWcRef?: HTMLKolButtonWcElement;

	private readonly catchRef = (ref?: HTMLKolButtonWcElement) => {
		this.buttonWcRef = ref;
	};

	/**
	 * @deprecated Use kolFocus instead.
	 */
	@Method()
	public async focus() {
		await this.kolFocus();
	}

	@Method()
	// eslint-disable-next-line @typescript-eslint/require-await
	public async kolFocus() {
		await this.buttonWcRef?.kolFocus();
	}

	private toggleTimeout?: ReturnType<typeof setTimeout>;

	private handleOnClick = (event: MouseEvent) => {
		this._open = !this._open;

		/**
		 * Der Timeout wird benötigt, damit das Event
		 * vom Button- auf das Accordion-Event wechselt.
		 * So ist es dem Anwendenden möglich das _open-
		 * Attribute abzufragen.
		 */

		clearTimeout(this.toggleTimeout);

		this.toggleTimeout = setTimeout(() => {
			if (this.host) {
				dispatchDomEvent(this.host, KolEvent.toggle, Boolean(this._open));
			}
			this.state._on?.onToggle?.(event, Boolean(this._open));
		}, 25);
	};

	public render(): JSX.Element {
		const { _open, _label, _disabled } = this.state;
		const _level = 1;

		const rootClass = 'kol-details';

		const props: CollapsibleProps = {
			id: this.nonce,
			label: _label,
			open: _open,
			disabled: _disabled,
			level: _level,
			onClick: this.handleOnClick,
			class: rootClass,
			HeadingProps: { class: `${rootClass}__heading` },
			HeadingButtonProps: {
				ref: this.catchRef,
				class: `${rootClass}__heading-button`,
				_icons: 'codicon codicon-chevron-right',
			},
			ContentProps: {
				class: `${rootClass}__content indented-text`,
				wrapperClass: `${rootClass}__wrapper`,
				animationClass: `${rootClass}__wrapper-animation`,
			},
		};

		return (
			<KolCollapsibleFc {...props}>
				<slot />
			</KolCollapsibleFc>
		);
	}

	/**
	 * Makes the element not focusable and ignore all events.
	 */
	@Prop() public _disabled?: boolean = false;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.).
	 */
	@Prop() public _label!: LabelPropType;

	/**
	 * Defines which H-level from 1-6 the heading has. 0 specifies no heading and is shown as bold text.
	 */
	@Prop() public _level?: HeadingLevel = 1;

	/**
	 * Defines the callback functions for details.
	 */
	@Prop() public _on?: DetailsCallbacksPropType<boolean>;

	/**
	 * If set (to true) opens/expands the element, closes if not set (or set to false).
	 * @TODO: Change type back to `OpenPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _open?: boolean = false;

	@State() public state: DetailsStates = {
		_label: '', // ⚠ required
		_level: 1,
		_on: {},
	};

	@Watch('_disabled')
	public validateDisabled(value?: DisabledPropType): void {
		validateDisabled(this, value);
	}

	@Watch('_label')
	public validateLabel(value?: LabelPropType): void {
		validateLabel(this, value, {
			required: true,
		});
	}

	@Watch('_level')
	public validateLevel(value?: HeadingLevel): void {
		watchHeadingLevel(this, value);
	}

	@Watch('_on')
	public validateOn(on?: DetailsCallbacksPropType<boolean>) {
		validateDetailsCallbacks(this, on);
	}

	@Watch('_open')
	public validateOpen(value?: boolean): void {
		validateOpen(this, value);
	}

	public componentWillLoad(): void {
		this.validateDisabled(this._disabled);
		this.validateLabel(this._label);
		this.validateLevel(this._level);
		this.validateOn(this._on);
		this.validateOpen(this._open);
	}
}
