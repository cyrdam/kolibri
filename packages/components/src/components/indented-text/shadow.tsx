import type { IndentedTextProps } from '../../schema';
import type { JSX } from '@stencil/core';
import { Component, h, Host } from '@stencil/core';
import { KolIndentedTextWcTag } from '../../core/component-names';

/**
 * @deprecated Will be removed in v3. Replace with custom HTML and styling.
 * @slot - Der Text.
 */
@Component({
	tag: 'kol-indented-text',
	styleUrls: {
		default: './style.scss',
	},
	shadow: true,
})
export class KolIndentedText implements IndentedTextProps {
	public render(): JSX.Element {
		return (
			<Host class="kol-indented-text">
				<KolIndentedTextWcTag>
					<slot />
				</KolIndentedTextWcTag>
			</Host>
		);
	}
}
