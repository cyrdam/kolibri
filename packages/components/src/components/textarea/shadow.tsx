import type { JSX } from '@stencil/core';
import { Component, Element, h, Method, Prop, State, Watch } from '@stencil/core';
import clsx from 'clsx';

import type {
	AdjustHeightPropType,
	CSSResize,
	FocusableElement,
	HasCounterPropType,
	HideErrorPropType,
	IdPropType,
	InputTypeOnDefault,
	KoliBriHorizontalIcons,
	LabelWithExpertSlotPropType,
	MsgPropType,
	NamePropType,
	RowsPropType,
	ShortKeyPropType,
	SpellCheckPropType,
	Stringified,
	SyncValueBySelectorPropType,
	TextareaAPI,
	TextareaStates,
	TooltipAlignPropType,
} from '../../schema';
import { setState } from '../../schema';

import { nonce } from '../../utils/dev.utils';
import KolFormFieldStateWrapperFc, { type FormFieldStateWrapperProps } from '../../functional-component-wrappers/FormFieldStateWrapper';
import KolTextAreaStateWrapperFc, { type TextAreaStateWrapperProps } from '../../functional-component-wrappers/TextAreaStateWrapper';
import KolInputContainerFc from '../../functional-component-wrappers/InputContainerStateWrapper';
import { TextareaController } from './controller';

/**
 * https://stackoverflow.com/questions/17772260/textarea-auto-height
 */
const increaseTextareaHeight = (el: HTMLTextAreaElement): number => {
	el.style.overflow = 'hidden'; // verhindert, dass ein Scrollbalken kurz angezeigt wird
	const currentRows = el.rows;
	const rowHeight = el.clientHeight / currentRows;
	el.rows = 1;
	const nextRows = Math.round(el.scrollHeight / rowHeight);
	el.rows = currentRows;
	return nextRows;
};

/**
 * @slot - Die Beschriftung des Eingabefeldes.
 */
@Component({
	tag: 'kol-textarea',
	styleUrls: {
		default: './style.scss',
	},
	shadow: {
		delegatesFocus: true,
	},
})
export class KolTextarea implements TextareaAPI, FocusableElement {
	@Element() private readonly host?: HTMLKolTextareaElement;
	private textareaRef?: HTMLTextAreaElement;

	private readonly catchRef = (ref?: HTMLTextAreaElement) => {
		this.textareaRef = ref;
	};

	@Method()
	// eslint-disable-next-line @typescript-eslint/require-await
	public async getValue(): Promise<string | undefined> {
		return this.textareaRef?.value;
	}

	@Method()
	// eslint-disable-next-line @typescript-eslint/require-await
	public async kolFocus() {
		this.textareaRef?.focus();
	}

	private getFormFieldProps(): FormFieldStateWrapperProps {
		return {
			state: this.state,
			class: clsx('kol-textarea', 'textarea', {
				'has-value': this.state._hasValue,
				'has-counter': !!this.state._hasCounter,
			}),
			tooltipAlign: this._tooltipAlign,
			onClick: () => this.textareaRef?.focus(),
			alert: this.showAsAlert(),
		};
	}

	private getTextAreaProps(): TextAreaStateWrapperProps {
		return {
			ref: this.catchRef,
			state: this.state,
			style: {
				resize: this.state._resize,
			},
			...this.controller.onFacade,
			onInput: this.onInput,
			onFocus: (event: Event) => {
				this.controller.onFacade.onFocus(event);
				this.inputHasFocus = true;
			},
			onBlur: (event: Event) => {
				this.controller.onFacade.onBlur(event);
				this.inputHasFocus = false;
			},
		};
	}

	public render(): JSX.Element {
		return (
			<KolFormFieldStateWrapperFc {...this.getFormFieldProps()}>
				<KolInputContainerFc state={this.state}>
					<KolTextAreaStateWrapperFc {...this.getTextAreaProps()} />
				</KolInputContainerFc>
			</KolFormFieldStateWrapperFc>
		);
	}

	private readonly controller: TextareaController;

	/**
	 * Defines which key combination can be used to trigger or focus the interactive element of the component.
	 */
	@Prop() public _accessKey?: string;

	/**
	 * Adjusts the height of the element to its content.
	 * @TODO: change back to AdjustHeightPropType after stencil #4663 has been resolved
	 */
	@Prop() public _adjustHeight?: boolean = false;

	/**
	 * Makes the element not focusable and ignore all events.
	 * @TODO: Change type back to `DisabledPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _disabled?: boolean = false;

	/**
	 * Shows the character count on the lower border of the input.
	 * @TODO: Change type back to `HasCounterPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hasCounter?: boolean = false;

	/**
	 * Hides the error message but leaves it in the DOM for the input's aria-describedby.
	 * @TODO: Change type back to `HideErrorPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _hideError?: boolean = false;

	/**
	 * Hides the caption by default and displays the caption text with a tooltip when the
	 * interactive element is focused or the mouse is over it.
	 * @TODO: Change type back to `HideLabelPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _hideLabel?: boolean = false;

	/**
	 * Defines the hint text.
	 */
	@Prop() public _hint?: string = '';

	/**
	 * Defines the icon classnames (e.g. `_icons="fa-solid fa-user"`).
	 */
	@Prop() public _icons?: KoliBriHorizontalIcons;

	/**
	 * Defines the internal ID of the primary component element.
	 */
	@Prop() public _id?: IdPropType;

	/**
	 * Defines the visible or semantic label of the component (e.g. aria-label, label, headline, caption, summary, etc.). Set to `false` to enable the expert slot.
	 */
	@Prop() public _label!: LabelWithExpertSlotPropType;

	/**
	 * Defines the maximum number of input characters.
	 */
	@Prop() public _maxLength?: number;

	/**
	 * Defines the properties for a message rendered as Alert component.
	 */
	@Prop() public _msg?: Stringified<MsgPropType>;

	/**
	 * Defines the technical name of an input field.
	 */
	@Prop() public _name?: NamePropType;

	/**
	 * Gibt die EventCallback-Funktionen für das Input-Event an.
	 */
	@Prop() public _on?: InputTypeOnDefault;

	/**
	 * Defines the placeholder for input field. To be shown when there's no value.
	 */
	@Prop() public _placeholder?: string;

	/**
	 * Makes the input element read only.
	 * @TODO: Change type back to `ReadOnlyPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _readOnly?: boolean = false;

	/**
	 * Defines whether and in which direction the size of the input can be changed by the user. (https://developer.mozilla.org/de/docs/Web/CSS/resize)
	 * In version 3 (v3), horizontal resizing is abolished. The corresponding property is then reduced to the properties `vertical` (default) and `none`.
	 */
	@Prop() public _resize?: CSSResize = 'vertical';

	/**
	 * Makes the input element required.
	 * @TODO: Change type back to `RequiredPropType` after Stencil#4663 has been resolved.
	 */
	@Prop() public _required?: boolean = false;

	/**
	 * Defines how many rows of text should be visible at the same time.
	 */
	@Prop({ mutable: true, reflect: false }) public _rows?: RowsPropType;

	/**
	 * Adds a visual short key hint to the component.
	 */
	@Prop() public _shortKey?: ShortKeyPropType;

	/**
	 * Defines whether the browser should check the spelling and grammar.
	 */
	@Prop() public _spellCheck?: SpellCheckPropType;

	/**
	 * Selector for synchronizing the value with another input element.
	 * @internal
	 */
	@Prop() public _syncValueBySelector?: SyncValueBySelectorPropType;

	/**
	 * Defines which tab-index the primary element of the component has. (https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/tabindex)
	 */
	@Prop() public _tabIndex?: number;

	/**
	 * Defines where to show the Tooltip preferably: top, right, bottom or left.
	 */
	@Prop() public _tooltipAlign?: TooltipAlignPropType = 'top';

	/**
	 * Shows if the input was touched by a user.
	 * @TODO: Change type back to `TouchedPropType` after Stencil#4663 has been resolved.
	 */
	@Prop({ mutable: true, reflect: true }) public _touched?: boolean = false;

	/**
	 * Defines the value of the input.
	 */
	@Prop() public _value?: string;

	@State() public state: TextareaStates = {
		_adjustHeight: false,
		_currentLength: 0,
		_hasValue: false,
		_hideError: false,
		_id: `id-${nonce()}`,
		_label: '', // ⚠ required
		_resize: 'vertical',
	};

	@State() private inputHasFocus = false;

	public constructor() {
		this.controller = new TextareaController(this, 'textarea', this.host);
	}

	private showAsAlert(): boolean {
		return Boolean(this.state._touched) && !this.inputHasFocus;
	}

	@Watch('_accessKey')
	public validateAccessKey(value?: string): void {
		this.controller.validateAccessKey(value);
	}

	@Watch('_adjustHeight')
	public validateAdjustHeight(value?: AdjustHeightPropType): void {
		this.controller.validateAdjustHeight(value);
	}

	@Watch('_disabled')
	public validateDisabled(value?: boolean): void {
		this.controller.validateDisabled(value);
	}

	@Watch('_hasCounter')
	public validateHasCounter(value?: HasCounterPropType): void {
		this.controller.validateHasCounter(value);
	}

	@Watch('_hideError')
	public validateHideError(value?: HideErrorPropType): void {
		this.controller.validateHideError(value);
	}

	@Watch('_hideLabel')
	public validateHideLabel(value?: boolean): void {
		this.controller.validateHideLabel(value);
	}

	@Watch('_hint')
	public validateHint(value?: string): void {
		this.controller.validateHint(value);
	}

	@Watch('_icons')
	public validateIcons(value?: Stringified<KoliBriHorizontalIcons>): void {
		this.controller.validateIcons(value);
	}

	@Watch('_id')
	public validateId(value?: string): void {
		this.controller.validateId(value);
	}

	@Watch('_label')
	public validateLabel(value?: LabelWithExpertSlotPropType): void {
		this.controller.validateLabel(value);
	}

	@Watch('_maxLength')
	public validateMaxLength(value?: number): void {
		this.controller.validateMaxLength(value);
	}

	@Watch('_msg')
	public validateMsg(value?: Stringified<MsgPropType>): void {
		this.controller.validateMsg(value);
	}

	@Watch('_name')
	public validateName(value?: string): void {
		this.controller.validateName(value);
	}

	@Watch('_on')
	public validateOn(value?: InputTypeOnDefault): void {
		this.controller.validateOn(value);
	}

	@Watch('_placeholder')
	public validatePlaceholder(value?: string): void {
		this.controller.validatePlaceholder(value);
	}

	@Watch('_readOnly')
	public validateReadOnly(value?: boolean): void {
		this.controller.validateReadOnly(value);
	}

	@Watch('_resize')
	public validateResize(value?: CSSResize): void {
		this.controller.validateResize(value);
	}

	@Watch('_required')
	public validateRequired(value?: boolean): void {
		this.controller.validateRequired(value);
	}

	@Watch('_rows')
	public validateRows(value?: RowsPropType): void {
		this.controller.validateRows(value);
	}

	@Watch('_shortKey')
	public validateShortKey(value?: ShortKeyPropType): void {
		this.controller.validateShortKey(value);
	}

	@Watch('_spellCheck')
	public validateSpellCheck(value?: SpellCheckPropType): void {
		this.controller.validateSpellCheck(value);
	}

	@Watch('_syncValueBySelector')
	public validateSyncValueBySelector(value?: SyncValueBySelectorPropType): void {
		this.controller.validateSyncValueBySelector(value);
	}

	@Watch('_tabIndex')
	public validateTabIndex(value?: number): void {
		this.controller.validateTabIndex(value);
	}

	@Watch('_touched')
	public validateTouched(value?: boolean): void {
		this.controller.validateTouched(value);
	}

	@Watch('_value')
	public validateValue(value?: string): void {
		this.controller.validateValue(value);
	}

	public componentDidLoad(): void {
		setTimeout(() => {
			if (this._adjustHeight === true && this.textareaRef /* SSR instanceof HTMLTextAreaElement */) {
				this._rows =
					this.state?._rows && this.state._rows > increaseTextareaHeight(this.textareaRef) ? this.state._rows : increaseTextareaHeight(this.textareaRef);
			} else if (!this._rows) {
				this._rows = 1;
			}
		}, 0);
	}

	public componentWillLoad(): void {
		this._touched = this._touched === true;
		this.controller.componentWillLoad();
		this.state._hasValue = !!this.state._value;
		this.controller.addValueChangeListener((v) => (this.state._hasValue = !!v));
	}

	private readonly onInput = (event: InputEvent) => {
		if (this.textareaRef instanceof HTMLTextAreaElement) {
			setState(this, '_currentLength', this.textareaRef.value.length);
			if (this.state._adjustHeight) {
				this._rows = increaseTextareaHeight(this.textareaRef);
			}
			this.controller.onFacade.onInput(event);
		}
	};
}
