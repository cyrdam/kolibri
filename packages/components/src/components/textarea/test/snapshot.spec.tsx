import { executeTests } from 'stencil-awesome-test';

import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';

import { getTextareaHtml } from './html.mock';

import type { TextareaProps } from '../../../schema';
import type { SpecPage } from '@stencil/core/testing';
import { KolTextarea } from '../shadow';

executeTests<TextareaProps>(
	'Textarea',
	async (props): Promise<SpecPage> => {
		const page = await newSpecPage({
			components: [KolTextarea],
			template: () => <kol-textarea {...props} />,
		});
		return page;
	},
	{
		_label: ['Texteingabe'],
		_disabled: [true, false],
		_rows: ['5'],
		_msg: [{ _type: 'error', _description: 'Es ist ein Fehler aufgetreten' }],
		_value: ['Kleiner Text im Eingabefeld'],
		_placeholder: ['Hier steht ein Platzhaltertext'],
		_hint: ['Hinweis'],
		_touched: [true, false],
	},
	getTextareaHtml,
	{
		execMode: 'default', // ready
		needTimers: true,
	},
);
