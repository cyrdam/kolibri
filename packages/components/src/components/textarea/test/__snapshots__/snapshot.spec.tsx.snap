// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _accessKey="V" 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          Label
          <span aria-hidden="true" class="access-key-hint">
            V
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea accesskey="V" autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _alert=true 1`] = `
<kol-textarea _alert="" _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _disabled=true 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--disabled kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container kol-input-container--disabled">
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea kol-textarea--disabled" disabled="" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _hint="Hint" 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea aria-describedby="id-nonce-hint" autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
      <span class="kol-form-field__hint" id="id-nonce-hint">
        Hint
      </span>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _icons={"left":{"icon":"codicon codicon-arrow-left"},"right":{"icon":"codicon codicon-arrow-right"}} 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__adornment kol-input-container__adornment--start">
            <kol-icon _icons="codicon codicon-codicon codicon-arrow-left" _label=""></kol-icon>
          </div>
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
          <div class="kol-input-container__adornment kol-input-container__adornment--end">
            <kol-icon _icons="codicon codicon-codicon codicon-arrow-right" _label=""></kol-icon>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _icons={"left":{"icon":"codicon codicon-arrow-left"}} 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__adornment kol-input-container__adornment--start">
            <kol-icon _icons="codicon codicon-codicon codicon-arrow-left" _label=""></kol-icon>
          </div>
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
          <div class="kol-input-container__adornment kol-input-container__adornment--end"></div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _icons={"right":{"icon":"codicon codicon-arrow-right"}} 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__adornment kol-input-container__adornment--start"></div>
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
          <div class="kol-input-container__adornment kol-input-container__adornment--end">
            <kol-icon _icons="codicon codicon-codicon codicon-arrow-right" _label=""></kol-icon>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _msg={"_type":"error","_description":"Es ist ein Fehler aufgetreten"} _hideError=true 1`] = `
<kol-textarea _hide-error="" _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field--hidden-error kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _msg={"_type":"error","_description":"Es ist ein Fehler aufgetreten"} 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _readOnly=true 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field--read-only kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea kol-textarea--readonly" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" readonly="" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _shortKey="S" 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          Label
          <span aria-hidden="true" class="access-key-hint">
            S
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" _touched=true 1`] = `
<kol-textarea _touched="" _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field--touched kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea kol-textarea--touched" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;

exports[`kol-textarea should render with _label="Label" _name="field" _placeholder="Hier steht ein Platzhaltertext" _rows=5 _spellCheck=true _value="Value" 1`] = `
<kol-textarea _value="Value">
  <template shadowrootmode="open">
    <div class="kol-form-field kol-form-field--has-value kol-form-field-textarea" role="presentation">
      <label class="kol-form-field__label" htmlfor="id-nonce" id="id-nonce-label">
        <span class="kol-form-field__label-text">
          <span>
            Label
          </span>
        </span>
      </label>
      <div class="kol-form-field__input">
        <div class="kol-input-container">
          <div class="kol-input-container__container"><textarea autocapitalize="off" autocorrect="off" class="kol-textarea" id="id-nonce" name="field" placeholder="Hier steht ein Platzhaltertext" rows="5" value="Value" style="resize: vertical;"></textarea>
          </div>
        </div>
      </div>
    </div>
  </template>
</kol-textarea>
`;
